### DISCUSSION TITLE
Make datasource UID editable in GUI

### DISCUSSION THREAD

**Discussion by XANi** [grafana/grafana#45230] [Data Sources]
Previously, as long as 2 grafana instances (say dev and prod, or dc1 and dc2) had identically named data sources it was relatively straightforward to either export/import dashboards between them manually or automate that.

That was hacky (it relied on 2 instances having same name for data source), and caused problems with renaming ( #1664 https://github.com/grafana/grafana/pull/33817  ) but *generally* worked.

But in recent version exported dashboard JSON now contains UID.

Now name of the data source is user-controllable but UID is not, which means only way for new instance to accept dashboards exported from old instance is either:

* set up provisioning with same UIDs (which is fine solution if you're already automated your deployments but otherwise "too late")
* editing database to match that UID (and as I understand break any existing dashboards in the database)
* editing dashboard JSONs to use legacy by-name method

It would be useful if there was an option to change the datasouce UID, either in web config (altho that would possibly require migrating existing ones), or to have it as option during export.

**Comment by sojuz151** [comment]
What about making the data source into a variable with the instance name filter machtich only one name?

**Comment by XANi** [comment]
Dunno. "Export for sharing externally" already *kinda* does 90% of the work here as it replaces UID with "${DS_INFLUXDB}" but that doesn't help if your instance has more than one influxdb database used.

**Comment by xkilian** [comment]
This is also a huge issue for us, as we have air-gapped networks that have similar dashboards, and we can no longer import/export dashboards between the environments. We need a solution.

**Reply by XANi** [reply]
@xkilian the "hack" is to just make sure all of them have same UID (either changing one in DB or deploying one via provisioning) but it's definitely less trivial solution

**Comment by rafaelpirolla** [comment]
PLEASE TELL ME IF THIS WILL BREAK ANY OTHER PART OF CONFIG

Example:
get the uid by checking the datasource in a json of a current dashboard

```
systemcl stop grafana-server
sqlite3 /var/lib/grafana/grafana.db
>select * from data_source;
1|1|1|prometheus|Prometheus|...|...|...|...|...|...|...|...|...|...|...|...|...|...|...|000000001
...
>update data_source set uid="<uid here>" where name="Prometheus";
```

schema:
```
>.schema
(...)
CREATE TABLE `data_source` (
`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
, `org_id` INTEGER NOT NULL
, `version` INTEGER NOT NULL
, `type` TEXT NOT NULL
, `name` TEXT NOT NULL
, `access` TEXT NOT NULL
, `url` TEXT NOT NULL
, `password` TEXT NULL
, `user` TEXT NULL
, `database` TEXT NULL
, `basic_auth` INTEGER NOT NULL
, `basic_auth_user` TEXT NULL
, `basic_auth_password` TEXT NULL
, `is_default` INTEGER NOT NULL
, `json_data` TEXT NULL
, `created` DATETIME NOT NULL
, `updated` DATETIME NOT NULL
, `with_credentials` INTEGER NOT NULL DEFAULT 0, `secure_json_data` TEXT NULL, `read_only` INTEGER NULL, `uid` TEXT NOT NULL DEFAULT 0);
(...)
```

**Reply by mrdsam** [reply]
Very good workaround, thanks.

**Comment by shaunmansell** [comment]
Anyone know exactly what Grafana version brought in the change to the JSON spec to now include datasource UIDs?

I'm looking for the highest version I could roll back to where I won't hit issues in our cross environment dashboard provisioning that has been impacted by UIDs in the dashboard JSON

Edit: Answering my own question. It came out in 8.3.0 via https://github.com/grafana/grafana/pull/33817

**Comment by laumarf3** [comment]
It would be great to have the possibility of not exporting the UID, but also the database name!

**Comment by youssefbenfarhat** [comment]
Hello, I'm facing the same error.

Did you find a solution for this ?

**Comment by XANi** [comment]
@youssefbenfarhat we just changed our datasources to have same UID between sites (via provisioning) but that's a half measure

**Reply by youssefbenfarhat** [reply]
What do you mean by  a half measure ?

**Reply by XANi** [reply]
Well, we have a bunch of same instance using mostly same graphs but serving different sites. Wouldn't work too well if environment wasn't so homogenous

**Comment by gplechuck** [comment]
Hey @youssefbenfarhat , @XANi 

Stumbled across this post this morning after encountering a similar issue provisioning a dashboard across sites (with different datasource UIDs) . 

An alternative workaround - 

* export the dashboard JSON, click 'View JSON' and copy that
* change any occurrences of the datasource reference to the older style format , eg 
```
"datasource": {
  "type": "prometheus",
  "uid": "WAwoBA8Gk"
},
```
to this - 
`"datasource": "prometheus",`

This allowed us to share dashboards between sites where the datasources were similarly named (eg prometheus), but had different UIDs.

Hope that helps .

**Reply by dkrut** [reply]
Thanks! It helps for me. I'm using influxdb

**Reply by santiago-salinas** [reply]
Thanks! Very useful

**Comment by mozai** [comment]
Ran into this today.  Made a new panel in the test env, exported the dashboard JSON to our helmchart to deploy across the enterprise.  Next morning, emails from confused users because of "datasource not found" because even though the datasource definition is identical (same helm chart) across all clusters, the generated UID is different between test and prod.  If I could enforce a uid string in a datasource definition, like we can with dashboard definitions, that might help.

**Comment by XANi** [comment]
The "fix" for it currently is provisioning datasources via JSON files rather than via UI and keeping consistent UID

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82962. Feel free to continue the discussion around this there. Thank you!
